File: backend/__tests__/pricing.test.ts
import {
  MEAL_PLAN_UPLIFT_PERCENT,
  baseRoomRateFor,
  mealUpliftFor,
  mealOffsetFor,
  computeStayBreakdown,
  gstRatePercentFor,
  GST_PERCENT_STANDARD,
  GST_PERCENT_LUXURY,
  GST_LUXURY_THRESHOLD_PER_ROOM_NIGHT,
} from '../src/lib/pricing'
import { seedProperties, seedRoomTypes } from '../src/data/seed'
import type { MealPlan } from '../src/types'

/**
 * The two rules the client set on 5 Aug 2026, locked down.
 *
 * Neither had any test coverage before: the invoice suite renders a PDF without
 * asserting a tax rate, and nothing anywhere asserted a meal price. Both numbers
 * are money the guest is charged, and both were wrong in production until this
 * change, so they get tests that fail loudly if anyone edits the constants.
 */

const PLANS: MealPlan[] = ['Room Only', 'With Breakfast', 'All Meals Included']

describe('GST — 5%, not 12% (client, 5 Aug 2026)', () => {
  it('taxes an ordinary room at 5%', () => {
    expect(GST_PERCENT_STANDARD).toBe(5)
    expect(gstRatePercentFor(2000)).toBe(5)
    expect(gstRatePercentFor(7499.99)).toBe(5)
  })

  it('keeps the 18% slab for supplies valued above the threshold', () => {
    expect(GST_PERCENT_LUXURY).toBe(18)
    expect(GST_LUXURY_THRESHOLD_PER_ROOM_NIGHT).toBe(7500)
    // The argument is GST-inclusive; the slab is defined on the value of supply.
    // 7,875 inclusive is exactly 7,500 net at the 5% rate, so the line sits there.
    expect(gstRatePercentFor(7875)).toBe(5)
    expect(gstRatePercentFor(7875.01)).toBe(18)
    expect(gstRatePercentFor(20000)).toBe(18)
  })

  it('does not push the ₹7,500 Royal Suite on All Meals into the 18% slab', () => {
    // Regression. EOK and Amar Inn Royal Suites are ₹5,000 + 50% = ₹7,500
    // inclusive, whose value of supply is ₹7,142.86 — inside the 5% band. The
    // old comparison read the inclusive ₹7,500 and over-charged 13 points of tax
    // on the group's two most expensive rooms.
    expect(gstRatePercentFor(7500)).toBe(5)
    expect(7500 / 1.05).toBeCloseTo(7142.86, 2)
  })

  /**
   * Which seeded rooms actually reach the upper slab.
   *
   * Until 1 Sep 2026 the answer was "none", and this test asserted exactly that.
   * Hotel Amaltas International changed it: its Superior is ₹6,000, and All
   * Meals adds 50%, so the quote is ₹9,000 inclusive — a value of supply of
   * ₹8,571, comfortably past the ₹7,500 line. It is the first room in the group
   * to cross it, and 18% on it is correct, not a regression.
   *
   * The assertion is therefore an exact set rather than a blanket "all 5%": it
   * still fails loudly if a rate change quietly pushes another room over, and it
   * fails just as loudly if the threshold logic starts over-taxing rooms that
   * belong in the 5% band — which is the bug the ₹7,500 Royal Suite test above
   * guards from the other side.
   */
  it('quotes every seeded room in the 5% slab except the one that genuinely clears the threshold', () => {
    const luxury: string[] = []

    for (const room of seedRoomTypes) {
      const prop = seedProperties.find((p) => p.id === room.property_id)!
      const base = baseRoomRateFor(prop.base_price, room.price_offset)
      for (const plan of PLANS) {
        const inclusive = base + mealUpliftFor(plan, base)
        if (gstRatePercentFor(inclusive) === GST_PERCENT_LUXURY) {
          luxury.push(`${prop.slug}/${room.slug}/${plan}`)
          // Never 18% on a supply that is actually inside the band.
          expect(inclusive / (1 + GST_PERCENT_STANDARD / 100))
            .toBeGreaterThan(GST_LUXURY_THRESHOLD_PER_ROOM_NIGHT)
        }
      }
    }

    expect(luxury).toEqual(['hotel-amaltas-international/superior-room/All Meals Included'])
  })

  it('puts the Amaltas Superior in the 5% slab on Room Only and Breakfast', () => {
    // Only the MAP quote crosses. ₹6,000 alone, and ₹7,500 with breakfast, are
    // both inside the band — the second by the same ₹7,142.86 margin as the
    // Royal Suite above, so the two sit on the threshold from opposite sides.
    expect(gstRatePercentFor(6000)).toBe(5)
    expect(gstRatePercentFor(7500)).toBe(5)
    expect(gstRatePercentFor(9000)).toBe(18)
  })

  it('splits a total inclusive of 5% into base + tax that add back up', () => {
    const total = 6000
    const rate = gstRatePercentFor(total)
    const base = Math.round((total / (1 + rate / 100)) * 100) / 100
    const tax = Math.round((total - base) * 100) / 100
    expect(rate).toBe(5)
    expect(base).toBeCloseTo(5714.29, 2)
    expect(tax).toBeCloseTo(285.71, 2)
    expect(Math.round((base + tax) * 100) / 100).toBe(total)
  })
})

describe('Meal plans — percentage of the base room rate (client, 5 Aug 2026)', () => {
  it('is EP 0 / CP 25 / MAP 50', () => {
    expect(MEAL_PLAN_UPLIFT_PERCENT['Room Only']).toBe(0)
    expect(MEAL_PLAN_UPLIFT_PERCENT['With Breakfast']).toBe(25)
    expect(MEAL_PLAN_UPLIFT_PERCENT['All Meals Included']).toBe(50)
  })

  it('measures the percentage off base price + category offset, not base price alone', () => {
    // Downtown EOK: ₹3,000 base, Super Deluxe +₹1,000 => ₹4,000.
    expect(baseRoomRateFor(3000, 1000)).toBe(4000)
    expect(mealUpliftFor('With Breakfast', 4000)).toBe(1000)
    expect(mealUpliftFor('All Meals Included', 4000)).toBe(2000)

    // If it were measured off the ₹3,000 base alone, CP on a Super Deluxe would
    // cost the same as CP on a Deluxe. It must not.
    expect(mealUpliftFor('With Breakfast', baseRoomRateFor(3000, 0)))
      .not.toBe(mealUpliftFor('With Breakfast', baseRoomRateFor(3000, 1000)))
  })

  it('makes CP exactly 1.25x and MAP exactly 1.5x the nightly rate, at every hotel', () => {
    for (const room of seedRoomTypes) {
      const prop = seedProperties.find((p) => p.id === room.property_id)!
      const base = baseRoomRateFor(prop.base_price, room.price_offset)

      const nightly = (plan: MealPlan) => base + mealUpliftFor(plan, base)

      expect(nightly('Room Only')).toBe(base)
      expect(nightly('With Breakfast')).toBe(base * 1.25)
      expect(nightly('All Meals Included')).toBe(base * 1.5)
    }
  })

  it('applies the same percentage at every property', () => {
    // The real invariant is that no two properties share a slug — `room_types.id`
    // is derived from it and a collision would cross two hotels' inventory. This
    // used to read `toBe(9)`, which asserted the size of the group instead and
    // failed the day a tenth hotel was added.
    const slugs = new Set(seedProperties.map((p) => p.slug))
    expect(slugs.size).toBe(seedProperties.length)

    for (const room of seedRoomTypes) {
      const prop = seedProperties.find((p) => p.id === room.property_id)!
      const base = baseRoomRateFor(prop.base_price, room.price_offset)
      expect((room.breakfast_offset / base) * 100).toBe(25)
      expect((room.all_meals_offset / base) * 100).toBe(50)
    }
  })

  it('resolves the same figure from a joined row, an explicit base rate and the seeded columns', () => {
    for (const room of seedRoomTypes) {
      const prop = seedProperties.find((p) => p.id === room.property_id)!
      const base = baseRoomRateFor(prop.base_price, room.price_offset)

      for (const plan of PLANS) {
        const fromJoin = mealOffsetFor(plan, { ...room, base_price: prop.base_price })
        const fromExplicit = mealOffsetFor(plan, room, base)
        const fromColumns = mealOffsetFor(plan, room) // in-memory: no base_price
        expect(fromJoin).toBe(fromExplicit)
        expect(fromColumns).toBe(fromExplicit)
      }
    }
  })
})

describe('Meal uplift composes with the other percentages without double-counting', () => {
  const base = 4000 // EOK Super Deluxe
  const stay = {
    basePrice: 3000,
    roomOffset: 1000,
    weekendSurchargePercent: 20,
    checkIn: '2026-09-07', // Mon
    checkOut: '2026-09-08',
    roomsCount: 1,
  }

  it('applies the meal percentage once, not to itself', () => {
    const total = computeStayBreakdown({
      ...stay,
      mealOffset: mealUpliftFor('All Meals Included', base),
    }).total
    expect(total).toBe(6000) // 4000 * 1.5, not 4000 * 1.5 * 1.5
  })

  it('lets the weekend surcharge run on the meal-inclusive rate', () => {
    const friday = computeStayBreakdown({
      ...stay,
      checkIn: '2026-09-11', // Fri
      checkOut: '2026-09-12',
      mealOffset: mealUpliftFor('All Meals Included', base),
    }).total
    expect(friday).toBe(7200) // 6000 * 1.2
  })

  it('charges the extra adult on the meal-inclusive rate, once per night per head', () => {
    const b = computeStayBreakdown({
      ...stay,
      mealOffset: mealUpliftFor('All Meals Included', base),
      extraAdults: 1,
      extraAdultPercent: 30,
    })
    expect(b.roomTotal).toBe(6000)
    expect(b.extraAdultTotal).toBe(1800) // 30% of 6000
    expect(b.total).toBe(7800)
  })

  it('does not multiply the extra bed by the room count', () => {
    const b = computeStayBreakdown({
      ...stay,
      roomsCount: 3,
      mealOffset: mealUpliftFor('With Breakfast', base),
      extraAdults: 1,
      extraAdultPercent: 30,
    })
    expect(b.roomTotal).toBe(15000) // 5000 * 3
    expect(b.extraAdultTotal).toBe(1500) // one extra bed, not three
  })

  it('charges nothing extra for Room Only', () => {
    expect(computeStayBreakdown({ ...stay, mealOffset: mealUpliftFor('Room Only', base) }).total)
      .toBe(4000)
  })
})

File: backend-go/internal/service/pricing/pricing_test.go
package pricing

import (
	"fmt"
	"math"
	"testing"

	"quadis-backend-go/internal/domain"
	"quadis-backend-go/internal/repository"
)

func TestGST_ThresholdAndRates(t *testing.T) {
	// Standard rate checks
	if rate := GSTRatePercentFor(2000); rate != 5 {
		t.Errorf("expected 5%% GST for 2000, got %f", rate)
	}
	if rate := GSTRatePercentFor(7499.99); rate != 5 {
		t.Errorf("expected 5%% GST for 7499.99, got %f", rate)
	}

	// 7875 inclusive = exactly 7500 net of 5% tax
	if rate := GSTRatePercentFor(7875); rate != 5 {
		t.Errorf("expected 5%% GST for 7875, got %f", rate)
	}
	if rate := GSTRatePercentFor(7875.01); rate != 18 {
		t.Errorf("expected 18%% GST for 7875.01, got %f", rate)
	}
	if rate := GSTRatePercentFor(20000); rate != 18 {
		t.Errorf("expected 18%% GST for 20000, got %f", rate)
	}

	// Royal Suite at 7500 inclusive must remain at 5%
	if rate := GSTRatePercentFor(7500); rate != 5 {
		t.Errorf("expected 5%% GST for 7500, got %f", rate)
	}
}

func TestGST_AmaltasSuperiorIsOnlyLuxury(t *testing.T) {
	plans := []domain.MealPlan{
		domain.MealPlanRoomOnly,
		domain.MealPlanWithBreakfast,
		domain.MealPlanAllMealsIncluded,
	}

	var luxuryList []string

	rooms := repository.BuildSeedRoomTypes()
	for _, room := range rooms {
		var prop *domain.PropertyRecord
		for _, p := range repository.SeedProperties {
			if p.ID == room.PropertyID {
				prop = &p
				break
			}
		}
		if prop == nil {
			continue
		}

		base := BaseRoomRateFor(prop.BasePrice, room.PriceOffset)
		for _, plan := range plans {
			inclusive := base + MealUpliftFor(plan, base)
			if GSTRatePercentFor(inclusive) == GSTPercentLuxury {
				luxuryList = append(luxuryList, fmt.Sprintf("%s/%s/%s", prop.Slug, room.Slug, plan))
			}
		}
	}

	expected := []string{"hotel-amaltas-international/superior-room/All Meals Included"}
	if len(luxuryList) != len(expected) {
		t.Fatalf("expected exactly %v, got %v", expected, luxuryList)
	}
	for i := range expected {
		if luxuryList[i] != expected[i] {
			t.Errorf("expected %s, got %s", expected[i], luxuryList[i])
		}
	}

	// Amaltas Superior on Room Only (6000) and Breakfast (7500) must be 5%
	if rate := GSTRatePercentFor(6000); rate != 5 {
		t.Errorf("expected 5%% for 6000, got %f", rate)
	}
	if rate := GSTRatePercentFor(7500); rate != 5 {
		t.Errorf("expected 5%% for 7500, got %f", rate)
	}
	if rate := GSTRatePercentFor(9000); rate != 18 {
		t.Errorf("expected 18%% for 9000, got %f", rate)
	}
}

func TestMealPlans_Uplifts(t *testing.T) {
	if MealPlanUpliftPercent(domain.MealPlanRoomOnly) != 0 {
		t.Errorf("expected 0 for Room Only")
	}
	if MealPlanUpliftPercent(domain.MealPlanWithBreakfast) != 25 {
		t.Errorf("expected 25 for With Breakfast")
	}
	if MealPlanUpliftPercent(domain.MealPlanAllMealsIncluded) != 50 {
		t.Errorf("expected 50 for All Meals Included")
	}

	// Downtown EOK: 3000 base + 1000 super deluxe = 4000
	base := BaseRoomRateFor(3000, 1000)
	if base != 4000 {
		t.Errorf("expected base 4000, got %f", base)
	}
	if uplift := MealUpliftFor(domain.MealPlanWithBreakfast, base); uplift != 1000 {
		t.Errorf("expected 1000 for CP, got %f", uplift)
	}
	if uplift := MealUpliftFor(domain.MealPlanAllMealsIncluded, base); uplift != 2000 {
		t.Errorf("expected 2000 for MAP, got %f", uplift)
	}
}

func TestStayPricing_Composition(t *testing.T) {
	base := 4000.0
	stay := StayPricingInput{
		BasePrice:               3000,
		RoomOffset:              1000,
		WeekendSurchargePercent: 20,
		CheckIn:                 "2026-09-07", // Mon
		CheckOut:                "2026-09-08", // Tue
		RoomsCount:              1,
	}

	// 1. All Meals Included on weekday: 4000 * 1.5 = 6000
	stay1 := stay
	stay1.MealOffset = MealUpliftFor(domain.MealPlanAllMealsIncluded, base)
	b1 := ComputeStayBreakdown(stay1)
	if b1.Total != 6000 {
		t.Errorf("expected total 6000, got %f", b1.Total)
	}

	// 2. Weekend surcharge (Friday) runs on meal-inclusive rate: 6000 * 1.2 = 7200
	stay2 := stay1
	stay2.CheckIn = "2026-09-11"  // Fri
	stay2.CheckOut = "2026-09-12" // Sat
	b2 := ComputeStayBreakdown(stay2)
	if b2.Total != 7200 {
		t.Errorf("expected total 7200 for Friday stay, got %f", b2.Total)
	}

	// 3. Extra adult charged on meal-inclusive rate: 6000 + (30% of 6000 = 1800) = 7800
	stay3 := stay1
	stay3.ExtraAdults = 1
	pAdult := 30.0
	stay3.ExtraAdultPercent = &pAdult
	b3 := ComputeStayBreakdown(stay3)
	if b3.RoomTotal != 6000 {
		t.Errorf("expected room total 6000, got %f", b3.RoomTotal)
	}
	if b3.ExtraAdultTotal != 1800 {
		t.Errorf("expected extra adult total 1800, got %f", b3.ExtraAdultTotal)
	}
	if b3.Total != 7800 {
		t.Errorf("expected total 7800, got %f", b3.Total)
	}

	// 4. Multiple rooms: 3 rooms with breakfast (5000 * 3 = 15000), 1 extra adult = 1500
	stay4 := stay
	stay4.RoomsCount = 3
	stay4.MealOffset = MealUpliftFor(domain.MealPlanWithBreakfast, base) // 1000 -> nightly = 5000
	stay4.ExtraAdults = 1
	stay4.ExtraAdultPercent = &pAdult
	b4 := ComputeStayBreakdown(stay4)
	if b4.RoomTotal != 15000 {
		t.Errorf("expected room total 15000, got %f", b4.RoomTotal)
	}
	if b4.ExtraAdultTotal != 1500 {
		t.Errorf("expected extra adult total 1500, got %f", b4.ExtraAdultTotal)
	}
	if b4.Total != 16500 {
		t.Errorf("expected total 16500, got %f", b4.Total)
	}
}

func TestStayPricing_NightOverrides(t *testing.T) {
	stay := StayPricingInput{
		BasePrice:               2500,
		RoomOffset:              0,
		MealOffset:              0,
		WeekendSurchargePercent: 0,
		CheckIn:                 "2026-10-01",
		CheckOut:                "2026-10-03", // 2 nights
		RoomsCount:              1,
		ExtraAdults:             1,
		NightOverrides: map[string]NightOverride{
			"2026-10-01": {
				Rate: 3500, // pushed rate
			},
		},
	}

	b := ComputeStayBreakdown(stay)
	// Night 1: 3500 + 30% extra adult (1050) = 4550
	// Night 2: 2500 + 30% extra adult (750) = 3250
	// Total room: 6000, Total extra adult: 1800 -> Total: 7800
	if b.RoomTotal != 6000 {
		t.Errorf("expected room total 6000, got %f", b.RoomTotal)
	}
	if b.ExtraAdultTotal != 1800 {
		t.Errorf("expected extra adult total 1800, got %f", b.ExtraAdultTotal)
	}
	if b.Total != 7800 {
		t.Errorf("expected total 7800, got %f", b.Total)
	}
}

func TestOccupancy_ChargeableGuests(t *testing.T) {
	tests := []struct {
		name          string
		adults        int
		childAges     []int
		roomsCount    int
		wantExtraAd   int
		wantExtraCh   int
	}{
		{"2 adults 1 room", 2, nil, 1, 0, 0},
		{"3 adults 1 room", 3, nil, 1, 1, 0},
		{"2 adults + 1 toddler (age 5)", 2, []int{5}, 1, 0, 0},
		{"2 adults + 1 concession child (age 10)", 2, []int{10}, 1, 0, 1},
		{"2 adults + 1 teen (age 14)", 2, []int{14}, 1, 1, 0},
		{"3 adults in 2 rooms", 3, nil, 2, 0, 0},
		{"4 adults + 2 children (age 9, 11) in 2 rooms", 4, []int{9, 11}, 2, 0, 2},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			res := ChargeableGuestsFor(OccupancyInput{
				Adults:     tt.adults,
				ChildAges:  tt.childAges,
				RoomsCount: tt.roomsCount,
			})
			if res.ExtraAdults != tt.wantExtraAd {
				t.Errorf("extra adults: got %d, want %d", res.ExtraAdults, tt.wantExtraAd)
			}
			if res.ExtraChildren != tt.wantExtraCh {
				t.Errorf("extra children: got %d, want %d", res.ExtraChildren, tt.wantExtraCh)
			}
		})
	}
}

func TestGST_ValueSplit(t *testing.T) {
	total := 6000.0
	rate := GSTRatePercentFor(total)
	base := math.Round((total/(1.0+rate/100.0))*100) / 100
	tax := math.Round((total-base)*100) / 100

	if rate != 5 {
		t.Errorf("expected 5%%, got %f", rate)
	}
	if math.Abs(base-5714.29) > 0.01 {
		t.Errorf("expected base ~5714.29, got %f", base)
	}
	if math.Abs(tax-285.71) > 0.01 {
		t.Errorf("expected tax ~285.71, got %f", tax)
	}
	if math.Round((base+tax)*100)/100 != total {
		t.Errorf("base + tax does not equal total")
	}
}
